src/plugins/arithmetics/add_operation.cpp

```diff
@@ -342,32 +389,302 @@ namespace phylanx { namespace execution_tree { namespace primitives
+            }
+            else
+            {
+                for (std::size_t i = 0; i != cm.rows(); ++i)
+                {
+                    blaze::row(cm, i) += blaze::trans(cv);
+                }
+
+                return primitive_argument_type{std::move(lhs)};
+            }
+        }
+        HPX_THROW_EXCEPTION(hpx::bad_parameter,
+            "add_operation::add2d1d",
+            execution_tree::generate_error_message(
+                "vector size does not match either the number of matrix "
+                "columns nor rows.",
+                name_, codename_));
+    }
+
+    int add_operation::get_stretch_dimension(
+        std::size_t lhs, std::size_t rhs) const
+    {
+        // 0: No copy, 1: stretch lhs, 2: stretch rhs
+        if (lhs != rhs)
+        {
+            // lhs x dimension must be stretched
+            if (lhs == 1)
+            {
+                return 1;
+            }
+            // rhs x dimensions must be stretched
+            else if (rhs == 1)
+            {
+                return 2;
+            }
         }
+        return 0;
+    }
 
-        // TODO: Blaze does not support broadcasting
+    primitive_argument_type add_operation::add2d2d_no_stretch(
+        arg_type&& lhs, arg_type&& rhs) const
+    {
+        // Avoid overwriting references, avoid memory reallocation when possible
         if (lhs.is_ref())
         {
-            blaze::DynamicMatrix<double> m{ cm.rows(), cm.columns() };
-            for (std::size_t i = 0; i != cm.rows(); ++i)
+            // Cannot reuse the memory if an operand is a reference
+            if (rhs.is_ref())
             {
-                blaze::row(m, i) = blaze::row(cm, i) + blaze::trans(cv);
+                lhs = lhs.matrix() + rhs.matrix();
+            }
+            // Reuse the memory from rhs operand
+            else
+            {
+                rhs.matrix() = lhs.matrix() + rhs.matrix();
             }
-            return primitive_argument_type{ std::move(m) };
+        }
+        // Reuse the memory from lhs operand
+        else
+        {
+            lhs.matrix() += rhs.matrix();
         }
 
-        for (std::size_t i = 0; i != cm.rows(); ++i)
+        return primitive_argument_type(std::move(lhs));
+    }
+
+    primitive_argument_type add_operation::add2d2d_stretch_x(
+        arg_type&& lhs, arg_type&& rhs, int stretch_x) const
+    {
+        // Stretch lhs x axis
+        if (stretch_x == 1)
         {
-            blaze::row(cm, i) += blaze::trans(cv);
+            // Ensure we can reuse memory
+            if (rhs.is_ref())
+            {
+                blaze::DynamicMatrix<double> result(
+                    rhs.dimension(0), rhs.dimension(1));
+
+                for (std::size_t i = 0; i != result.rows(); ++i)
+                {
+                    blaze::row(result, i) =
+                        blaze::row(lhs.matrix(), 0) +
+                        blaze::row(rhs.matrix(), i);
```

Comment:
You might want to avoid calling `.matrix()` for each row/column. Same applies below.